### PR TITLE
[VPlan] Add VPWidenInduction recipe as common base class (NFC).

### PR DESCRIPTION
**Pull request by fhahn**
This helps to simplify some existing code and new code
(https://github.com/llvm/llvm-project/pull/112145)

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -10131,13 +10129,7 @@ preparePlanForEpilogueVectorLoop(VPlan &Plan, Loop *L,
...
       // Retrieve the induction resume values for wide inductions from
       // their original phi nodes in the scalar loop.
-      PHINode *IndPhi = nullptr;
-      if (auto *Ind = dyn_cast<VPWidenPointerInductionRecipe>(&R)) {
-        IndPhi = cast<PHINode>(Ind->getUnderlyingValue());
-      } else {
-        auto *WidenInd = cast<VPWidenIntOrFpInductionRecipe>(&R);
-        IndPhi = WidenInd->getPHINode();
-      }
+      PHINode *IndPhi = dyn_cast<VPWidenInductionRecipe>(&R)->getPHINode();
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:10132]:
```suggestion
      PHINode *IndPhi = cast<VPWidenInductionRecipe>(&R)->getPHINode();
```

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 2]
```diff
@@ -2099,38 +2099,81 @@ class VPHeaderPHIRecipe : public VPSingleDefRecipe {
   }
 };
 
+/// Base class for widened induction (VPWidenIntOrFpInductionRecipe and
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2102]: Independent: perhaps worth handling Int and FP separately?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -1755,13 +1756,12 @@ void VPWidenIntOrFpInductionRecipe::execute(VPTransformState &State) {
 void VPWidenIntOrFpInductionRecipe::print(raw_ostream &O, const Twine &Indent,
                                           VPSlotTracker &SlotTracker) const {
   O << Indent << "WIDEN-INDUCTION";
-  if (getTruncInst()) {
+  if (auto *TI = getTruncInst()) {
     O << "\\l\"";
-    O << " +\n" << Indent << "\"  " << VPlanIngredient(IV) << "\\l\"";
-    O << " +\n" << Indent << "\"  ";
-    getVPValue(0)->printAsOperand(O, SlotTracker);
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1762]: Commit separately along with its test change?

File: llvm/test/Transforms/LoopVectorize/vplan-printing.ll [hunk 4]
```diff
@@ -655,10 +655,10 @@ define void @print_expand_scev(i64 %y, ptr %ptr) {
 ; CHECK-NEXT:    EMIT vp<[[CAN_IV:%.+]]> = CANONICAL-INDUCTION ir<0>, vp<[[CAN_IV_NEXT:%.+]]>
 ; CHECK-NEXT:     WIDEN-INDUCTION\l" +
 ; CHECK-NEXT:     "  %iv = phi %iv.next, 0\l" +
-; CHECK-NEXT:     "  ir<%v2>, vp<[[EXP_SCEV]]>, vp<[[VF]]>
+; CHECK-NEXT:     "  %v2 = trunc %iv, vp<[[EXP_SCEV]]>, vp<[[VF]]>
 ; CHECK-NEXT:     vp<[[DERIVED_IV:%.+]]> = DERIVED-IV ir<0> + vp<[[CAN_IV]]> * vp<[[EXP_SCEV]]>
 ; CHECK-NEXT:     vp<[[STEPS:%.+]]> = SCALAR-STEPS vp<[[DERIVED_IV]]>, vp<[[EXP_SCEV]]>
-; CHECK-NEXT:     WIDEN ir<%v3> = add nuw ir<%v2>, ir<1>
+; CHECK-NEXT:     WIDEN ir<%v3> = add nuw ir<%iv>, ir<1>
```
Comment (ayalz) [inline llvm/test/Transforms/LoopVectorize/vplan-printing.ll:661]: How/is this related? It originally adds %v2.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Done, thanks!

File: llvm/test/Transforms/LoopVectorize/vplan-printing.ll (hunk 4 above)
Comment (fhahn) [reply llvm/test/Transforms/LoopVectorize/vplan-printing.ll:659]: The patch changes `VPWidenIntOrFpInductionRecipe` to always use the phi node as underlying value (for consistency and to make it easier to retrieve it generically), whereas before the Trunc would be the underlying value.

This wasn't really obvious from the old printing for VPWidenIntOrFpInductionRecipe, which didn't show the result VPValue and VPValue operands properly. Modernized printing in 43045051d4114f2490bf0e6b01a7969d5c27ee04, hopefully clearer now and more in line with printing for other recipes.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Modernized printing for the recipe separately in 43045051d4114f2490bf0e6b01a7969d5c27ee04

File: llvm/test/Transforms/LoopVectorize/vplan-printing.ll (hunk 4 above)
Comment (ayalz) [reply llvm/test/Transforms/LoopVectorize/vplan-printing.ll:659]: Yeah, the switch to hold Trunc rather than IV separately is clear. Could Trunc continue to be printed as the result %ir<%v2> of the widened induction even if its no longer the underlying?

**Review by ayalz — APPROVED** [review]
LGTM, thanks for following up on this!

File: llvm/test/Transforms/LoopVectorize/vplan-printing.ll (hunk 4 above)
Comment (fhahn) [reply llvm/test/Transforms/LoopVectorize/vplan-printing.ll:659]: Unfortunately I don't think this would be possible, as at other places we usually print the underlying value of the VPValue, and trying to retrieve the Trunc would add a number of complexities.

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 5]
```diff
@@ -2099,38 +2099,81 @@ class VPHeaderPHIRecipe : public VPSingleDefRecipe {
...
   }
 
   VPWidenIntOrFpInductionRecipe(PHINode *IV, VPValue *Start, VPValue *Step,
                                 VPValue *VF, const InductionDescriptor &IndDesc,
                                 TruncInst *Trunc, DebugLoc DL)
-      : VPHeaderPHIRecipe(VPDef::VPWidenIntOrFpInductionSC, Trunc, Start, DL),
-        IV(IV), Trunc(Trunc), IndDesc(IndDesc) {
-    addOperand(Step);
+      : VPWidenInductionRecipe(VPDef::VPWidenIntOrFpInductionSC, IV, Start,
+                               Step, IndDesc, DL),
```
Comment (arcbbb) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2166]: This is not NFC because the underlying instruction changed from Trunc to IV here.
